File: Data Structures & Algorithms/house-robber-ii/submission-1.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int[][] dp = new int[2][nums.length];
        dp[0][0] = nums[0];
        dp[1][0] = 0;
        dp[0][1] = nums[0];
        dp[1][1] = nums[1];
        for (int i = 2; i < nums.length; i++) {
            dp[0][i] = Math.max(dp[0][i - 2] + nums[i], dp[0][i - 1]);
            dp[1][i] = Math.max(dp[1][i - 2] + nums[i], dp[1][i - 1]);
        }

        return Math.max(dp[0][nums.length - 2], dp[1][nums.length - 1]);
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = [1 for i in range(len(nums))] # initialize output to be all 1s
        
        for i in range(1, len(nums)):
            # set output[i] to its prefix product
            output[i] *= output[i - 1] * nums[i - 1]
        
        suffix = 1
        for i in range(len(nums) - 2, -1, -1):
            suffix *= nums[i + 1]
            output[i] *= suffix
        
        return output
            